replace incorrect Method.deleted_world with more useful Method.dispatch_status enum

vtjnash opened the pull request:
The original purpose of this was to manage quickly detecting if a method was replaced, but that stopped being correct after #53415. It is also a fairly heavy-weight description of that single bit of information, and apparently (despite intentionally not having "min" or "max" anywhere in the name) its existence misleads people into thinking that primary/deleted are somehow corresponding to min/max. That has always been meant to be a false equivalence: this bit of data only exists for invalidations, and primary_world is primarily valid to use for supporting generated functions (or cases like `show` where correctness isn't especially important, though we may use it at times as a short-cut to computing the result of a dispatch result).

~~This PR implements METHOD_ISMINMAX_INVOKE_LATEST fully (for more correctness), and provides just a stub for adding METHOD_ISMINMAX_CALL_LATEST (for more performance) later.~~

Also fixes #58215

vtjnash committed:
replace incorrect Method.deleted_world with more useful Method.dispatch_status enum

The original purpose of this was to manage quickly detecting if a method
was replaced, but that stopped being correct after #53415. It is also a
fairly heavy-weight description of a single bit of information, and
apparently (despite not having "min" or "max" anywhere in the name), its
existence misleads people into thinking that primary/deleted are somehow
corresponding to min/max. That is always false: this bit only exists for
invalidations, and primary_world is primarily valid to use for
supporting generated functions (or cases like `show` where correctness
isn't especially important, though we may use it at times as a short-cut
to computing the result of a dispatch result).

This implements METHOD_ISMINMAX_INVOKE_LATEST fully, and provides just a
stub for adding METHOD_ISMINMAX_CALL_LATEST later.

Fix #58215

Keno commented on src/julia_internal.h:
```diff
@@ -676,6 +676,9 @@ typedef union {
 #define SOURCE_MODE_NOT_REQUIRED            0x0
 #define SOURCE_MODE_ABI                     0x1
 
+#define METHOD_ISMINMAX_INVOKE_LATEST       0b0001
```
Needs documentation on what these mean.

JeffBezanson replied:
I think it could use a different name as well, as IIUC it's not really related to min-max world or invokelatest. Just a name collision to avoid.

topolarity commented on base/staticdata.jl:
```diff
@@ -340,24 +340,22 @@ function verify_call(@nospecialize(sig), expecteds::Core.SimpleVector, i::Int, n
 end
 
+const METHOD_ISMINMAX_INVOKE_LATEST = 0x1
+const METHOD_ISMINMAX_CALL_LATEST = 0x2
```
which calls / queries is this quantifying over exactly?

If it's "any call w/ sig <: method.sig" then it seems that this bit just represents whether the Method is more-specific than all intersecting Methods. Is that the right intuition?

vtjnash committed:
finish implementing METHOD_ISMINMAX_CALL_LATEST

This implements the equivalent performance optimization
METHOD_ISMINMAX_CALL_LATEST, though we should still teach
ml_matches_visitor / ml_matches fully-covering to use this info also to
skip further work once fully_covers is set on a method with this bit
also set (in any world).

Also rename:
METHOD_ISMINMAX_INVOKE_LATEST => METHOD_SIG_LATEST_WHICH
METHOD_ISMINMAX_CALL_LATEST => METHOD_SIG_LATEST_ONLY

vtjnash commented:
In addition to the name change, this now adds some of the call/load optimizations too (sort of like a 1-bit version of #58040)

```julia
$ hyperfine --show-output -- "./julia -e 'using Pkg, REPL'"
Benchmark 1: ./julia -e 'using Pkg, REPL'
  Time (mean ± σ):     865.7 ms ±   7.8 ms    [User: 675.5 ms, System: 191.9 ms]
  Range (min … max):   852.4 ms … 878.9 ms    10 runs

$ hyperfine --show-output -- "julia +1.12 -e 'using Pkg, REPL'"
Benchmark 1: julia +1.12 -e 'using Pkg, REPL'
  Time (mean ± σ):      1.052 s ±  0.015 s    [User: 0.827 s, System: 0.226 s]
  Range (min … max):    1.035 s …  1.079 s    10 runs

$ hyperfine --show-output -- "julia +v1.11 -e 'using Pkg, REPL'"
Benchmark 1: ./julia -e 'using Pkg, REPL'
  Time (mean ± σ):      1.002 s ±  0.027 s    [User: 0.761 s, System: 0.239 s]
  Range (min … max):    0.950 s …  1.042 s    10 runs
```

*Take these numbers with a heap of salt, since of my 4 terminal windows open, 2 of them run this PR about 20% slower than the other two, and those same 2 that run this PR faster also run v1.12 about 20% slower. There is otherwise no difference between those tmux panes other than that they consistently run hyperfine at different speeds.*

topolarity commented on base/staticdata.jl:
```diff
@@ -340,24 +364,25 @@ function verify_call(@nospecialize(sig), expecteds::Core.SimpleVector, i::Int, n
+# true indicates this method would be returned as the result from `which` when invoking `method.sig` in the current latest world
+const METHOD_SIG_LATEST_WHICH = 0x1
+# true indicates this method would be returned as the only result from `methods` when calling `method.sig` in the current latest world
+const METHOD_SIG_LATEST_ONLY = 0x2
```
These definitions are much easier to understand, thanks!

vtjnash committed:
bugfixes
